File: plugins/store/android/src/main/java/StorePlugin.kt
// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

package app.tauri.store

import android.app.Activity
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.Plugin
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.File

@TauriPlugin
class StorePlugin(private val activity: Activity) : Plugin(activity) {
    @Command
    fun load(invoke: Invoke) {
        try {
            val path = invoke.parseArgs(String::class.java)
            val file = File(activity.applicationContext.getExternalFilesDir(null), path)

            invoke.resolveObject(ObjectMapper().readTree(file))
        } catch (ex: Exception) {
            invoke.reject(ex.message)
        }
    }

    @Command
    fun save(invoke: Invoke) {
        try {
            val args = invoke.parseArgs(JsonNode::class.java)
            val path = args.get("store").asText()
            val cache = args.get("cache")
            val file = File(activity.applicationContext.getExternalFilesDir(null), path)

            if (!file.exists()) {
                file.parentFile?.mkdirs()
                file.createNewFile()
            }

            file.writeText(cache.toString())

            invoke.resolve()
        } catch (ex: Exception) {
            invoke.reject(ex.message)
        }
    }
}

File: plugins/shell/ios/Sources/StorePlugin.swift
// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

import Foundation

import SwiftRs
import Tauri
import UIKit
import WebKit


struct SaveStore: Codable {
    let store: String
    let cache: [String: JSON]
}

class StorePlugin: Plugin {

    @objc public func load(_ invoke: Invoke) throws {
        do {
            let urlString = try invoke.parseArgs(String.self)
            let url = URL(string: urlString)
            UIApplication.shared.openURL(url)
            invoke.resolve()
        } catch {
            invoke.reject(error.localizedDescription)
        }
    }
}


@_cdecl("init_plugin_shell")
func initPlugin() -> Plugin {
    return ShellPlugin()
}
